pkg/controller/nodemaintenance/drain.go

```diff
@@ -0,0 +1,187 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/kubectl/drain"
+)
+
+func runCordonOrUncordon(r *ReconcileNodeMaintenance, node *corev1.Node, desired bool) error {
+	cordonOrUncordon := "cordon"
+	if !desired {
+		cordonOrUncordon = "un" + cordonOrUncordon
+	}
+
+	log.Info(fmt.Sprintf("%s Node: %s", cordonOrUncordon, node.Name))
+
+	c := drain.NewCordonHelper(node)
+	if updateRequired := c.UpdateIfRequired(desired); updateRequired {
+		err, patchErr := c.PatchOrReplace(r.drainer.Client)
+		if patchErr != nil {
+			log.Error(err, fmt.Sprintf("Unable to %s Node %s \n", cordonOrUncordon, node.Name))
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func drainPods(r *ReconcileNodeMaintenance, nodeName string) error {
+	list, errs := r.drainer.GetPodsForDeletion(nodeName)
+
+	if errs != nil {
+		return utilerrors.NewAggregate(errs)
+	}
+
+	if warnings := list.Warnings(); warnings != "" {
+		log.Info(fmt.Sprintf("WARNING: %s\n", warnings))
+	}
+
+	if err := deleteOrEvictPods(r, list.Pods()); err != nil {
+		pendingList, newErrs := r.drainer.GetPodsForDeletion(nodeName)
+		log.Error(err, fmt.Sprintf("There are pending pods in node %q when an error occurred: \n", nodeName))
+
+		for _, pendingPod := range pendingList.Pods() {
+			log.Error(err, fmt.Sprintf("%s/%s\n", "pod", pendingPod.Name))
+		}
+		if newErrs != nil {
+			log.Error(err, fmt.Sprintf("following errors also occurred:\n%s", utilerrors.NewAggregate(newErrs)))
+		}
+		return err
+	}
+	return nil
+}
+
+// deleteOrEvictPods deletes or evicts the pods on the api server
+func deleteOrEvictPods(r *ReconcileNodeMaintenance, pods []corev1.Pod) error {
+	if len(pods) == 0 {
+		return nil
+	}
+
+	policyGroupVersion, err := drain.CheckEvictionSupport(r.drainer.Client)
+	if err != nil {
+		return err
+	}
+
+	getPodFn := func(namespace, name string) (*corev1.Pod, error) {
+		return r.drainer.Client.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
+	}
+
+	if len(policyGroupVersion) > 0 {
+		return evictPods(r, pods, policyGroupVersion, getPodFn)
+	}
+	return deletePods(r, pods, getPodFn)
+
+}
+
+func deletePods(r *ReconcileNodeMaintenance, pods []corev1.Pod, getPodFn func(namespace, name string) (*corev1.Pod, error)) error {
+	// 0 timeout means infinite, we use MaxInt64 to represent it.
+	var globalTimeout time.Duration
+	if r.drainer.Timeout == 0 {
+		globalTimeout = time.Duration(math.MaxInt64)
+	} else {
+		globalTimeout = r.drainer.Timeout
+	}
+	for _, pod := range pods {
+		err := r.drainer.DeletePod(pod)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+	_, err := waitForDelete(pods, 1*time.Second, globalTimeout, false, getPodFn)
+	return err
+}
+
+func evictPods(r *ReconcileNodeMaintenance, pods []corev1.Pod, policyGroupVersion string, getPodFn func(namespace, name string) (*corev1.Pod, error)) error {
+	returnCh := make(chan error, 1)
+
+	for _, pod := range pods {
+		go func(pod corev1.Pod, returnCh chan error) {
```

Comment:
I know this is coming from kubectl directly, but this looks like it has the potential to create a lot of traffic is more than one node with many pods are put into maintenance. If you run evict on three or four nodes in parallel and there is a non-standard situation, you have a good chance to do 1000 requests per second and either create a lot of load or get rate-limited.

Here we have a go-routine which does the delete for every pod in parallel + periodic polls on the pod state per pod in parallel. 

Does it have more than one worker on the controller?

Reply:
we have one worker